File: cpu/C#/src/StreamWorker.cs
using System;
using System.Runtime.InteropServices;
using System.Threading;
using System.Threading.Tasks;
using Avalonia;
using Avalonia.Media.Imaging;
using Avalonia.Platform;
using Avalonia.Threading;
using FFmpeg.AutoGen.Abstractions;

namespace RtspStressTest;

public sealed unsafe class StreamWorker : IDisposable
{
    private readonly int _streamId;
    private readonly string _rtspUrl;
    private readonly CancellationTokenSource _cts = new();

    private volatile bool _isConnected;
    private long _decodedFrames;
    private long _paintedFrames;
    private int _width = 2560;
    private int _height = 1440;
    private int _renderPending;
    private volatile bool _hasNewFrame;

    private WriteableBitmap? _writeableBitmap;

    private readonly int _targetWidth;
    private readonly int _targetHeight;

    private int _bufferWidth;
    private int _bufferHeight;

    private SwsContext* _swsContext;
    private int _swsSrcWidth;
    private int _swsSrcHeight;
    private int _swsDstWidth;
    private int _swsDstHeight;
    private AVPixelFormat _swsFormat = AVPixelFormat.AV_PIX_FMT_NONE;

    // Reusable arrays for sws_scale to eliminate GC allocations
    private readonly byte*[] _srcData = new byte*[8];
    private readonly int[] _srcStride = new int[8];
    private readonly byte*[] _dstData = new byte*[8];
    private readonly int[] _dstStride = new int[8];

    private Task? _workerTask;

    private long _currentPts = -1;
    private long _lastPresentedTimestamp;
    private double _lastDeltaMs;

    public int StreamId => _streamId;
    public string RtspUrl => _rtspUrl;
    public bool IsConnected => _isConnected;
    public ulong DecodedFrames => (ulong)Interlocked.Read(ref _decodedFrames);
    public ulong PaintedFrames => (ulong)Interlocked.Read(ref _paintedFrames);
    public long CurrentPts => Interlocked.Read(ref _currentPts);
    public double LastDeltaMs => _lastDeltaMs;
    public uint CurrentFps { get; set; }
    public uint CurrentPaintedFps { get; set; }
    public uint CurrentDecodedFps { get; set; }
    public int Width => _width;
    public int Height => _height;
    public WriteableBitmap? Bitmap => _writeableBitmap;

    public void IncrementPaintedFrames() => RecordPresentedFrame(CurrentPts);

    public void RecordPresentedFrame(long pts)
    {
        var now = System.Diagnostics.Stopwatch.GetTimestamp();
        var prev = Interlocked.Exchange(ref _lastPresentedTimestamp, now);
        if (prev > 0)
        {
            var deltaSec = (double)(now - prev) / System.Diagnostics.Stopwatch.Frequency;
            _lastDeltaMs = deltaSec * 1000.0;
        }
        Interlocked.Increment(ref _paintedFrames);
    }

    public bool HasNewFrame => _hasNewFrame;
    public void ResetNewFrame() => _hasNewFrame = false;

    public event Action? FrameRendered;

    public StreamWorker(int streamId, string rtspUrl, int targetWidth = 0, int targetHeight = 0)
    {
        _streamId = streamId;
        _rtspUrl = rtspUrl;
        _targetWidth = targetWidth;
        _targetHeight = targetHeight;

        var initialW = targetWidth > 0 ? targetWidth : _width;
        var initialH = targetHeight > 0 ? targetHeight : _height;

        // Initialize WriteableBitmap and pre-allocated buffer
        EnsureBuffers(initialW, initialH);
    }

    public void Start()
    {
        _workerTask = Task.Factory.StartNew(WorkerLoop, TaskCreationOptions.LongRunning);
    }

    private void EnsureBuffers(int width, int height)
    {
        if (_writeableBitmap == null || _bufferWidth != width || _bufferHeight != height)
        {
            _bufferWidth = width;
            _bufferHeight = height;

            var size = new PixelSize(width, height);
            _writeableBitmap = new WriteableBitmap(
                size,
                new Vector(96, 96),
                PixelFormat.Rgba8888,
                AlphaFormat.Opaque
            );
        }
    }

    private void EnsureSwsContext(int srcW, int srcH, AVPixelFormat format, int dstW, int dstH)
    {
        if (_swsContext == null || _swsSrcWidth != srcW || _swsSrcHeight != srcH ||
            _swsDstWidth != dstW || _swsDstHeight != dstH || _swsFormat != format)
        {
            if (_swsContext != null)
            {
                ffmpeg.sws_freeContext(_swsContext);
                _swsContext = null;
            }

            _swsContext = ffmpeg.sws_getContext(
                srcW,
                srcH,
                format,
                dstW,
                dstH,
                AVPixelFormat.AV_PIX_FMT_RGBA,
                (int)SwsFlags.SWS_FAST_BILINEAR,
                null,
                null,
                null
            );

            _swsSrcWidth = srcW;
            _swsSrcHeight = srcH;
            _swsDstWidth = dstW;
            _swsDstHeight = dstH;
            _swsFormat = format;
        }
    }

    private void WorkerLoop()
    {
        var pkt = ffmpeg.av_packet_alloc();
        var frame = ffmpeg.av_frame_alloc();

        while (!_cts.IsCancellationRequested)
        {
            AVFormatContext* fmtCtx = ffmpeg.avformat_alloc_context();
            if (fmtCtx == null)
            {
                Thread.Sleep(500);
                continue;
            }

            AVDictionary* opts = null;
            ffmpeg.av_dict_set(&opts, "rtsp_transport", "tcp", 0);
            ffmpeg.av_dict_set(&opts, "stimeout", "5000000", 0);       // 5 sec timeout (in us)
            ffmpeg.av_dict_set(&opts, "max_delay", "500000", 0);        // 500 ms max latency (in us)
            ffmpeg.av_dict_set(&opts, "buffer_size", "4194304", 0);     // 4MB socket buffer

            var ret = ffmpeg.avformat_open_input(&fmtCtx, _rtspUrl, null, &opts);
            ffmpeg.av_dict_free(&opts);

            if (ret < 0)
            {
                _isConnected = false;
                ffmpeg.avformat_close_input(&fmtCtx);
                if (!_cts.IsCancellationRequested)
                {
                    Thread.Sleep(3000);
                }
                continue;
            }

            if (ffmpeg.avformat_find_stream_info(fmtCtx, null) < 0)
            {
                _isConnected = false;
                ffmpeg.avformat_close_input(&fmtCtx);
                if (!_cts.IsCancellationRequested)
                {
                    Thread.Sleep(3000);
                }
                continue;
            }

            var videoStreamIdx = ffmpeg.av_find_best_stream(fmtCtx, AVMediaType.AVMEDIA_TYPE_VIDEO, -1, -1, null, 0);
            if (videoStreamIdx < 0)
            {
                _isConnected = false;
                ffmpeg.avformat_close_input(&fmtCtx);
                if (!_cts.IsCancellationRequested)
                {
                    Thread.Sleep(3000);
                }
                continue;
            }

            var codecPar = fmtCtx->streams[videoStreamIdx]->codecpar;
            // Pure CPU software decoding: avcodec_find_decoder
            var codec = ffmpeg.avcodec_find_decoder(codecPar->codec_id);
            if (codec == null)
            {
                Console.Error.WriteLine($"[Stream {_streamId}] Codec not found for ID: {codecPar->codec_id}");
                _isConnected = false;
                ffmpeg.avformat_close_input(&fmtCtx);
                if (!_cts.IsCancellationRequested)
                {
                    Thread.Sleep(3000);
                }
                continue;
            }

            var codecCtx = ffmpeg.avcodec_alloc_context3(codec);
            if (codecCtx == null)
            {
                _isConnected = false;
                ffmpeg.avformat_close_input(&fmtCtx);
                if (!_cts.IsCancellationRequested)
                {
                    Thread.Sleep(3000);
                }
                continue;
            }

            if (ffmpeg.avcodec_parameters_to_context(codecCtx, codecPar) < 0)
            {
                _isConnected = false;
                ffmpeg.avcodec_free_context(&codecCtx);
                ffmpeg.avformat_close_input(&fmtCtx);
                if (!_cts.IsCancellationRequested)
                {
                    Thread.Sleep(3000);
                }
                continue;
            }

            // Software decode constraints & thread tuning
            codecCtx->thread_count = 1; // 1 thread per decoder worker for balanced CPU core distribution
            codecCtx->flags |= ffmpeg.AV_CODEC_FLAG_LOW_DELAY;
            codecCtx->flags2 |= ffmpeg.AV_CODEC_FLAG2_FAST;

            if (ffmpeg.avcodec_open2(codecCtx, codec, null) < 0)
            {
                _isConnected = false;
                ffmpeg.avcodec_free_context(&codecCtx);
                ffmpeg.avformat_close_input(&fmtCtx);
                if (!_cts.IsCancellationRequested)
                {
                    Thread.Sleep(3000);
                }
                continue;
            }

            _isConnected = true;

            // Demux and decode loop
            while (!_cts.IsCancellationRequested)
            {
                ret = ffmpeg.av_read_frame(fmtCtx, pkt);
                if (ret < 0)
                {
                    // Connection dropped or EOF
                    break;
                }

                if (pkt->stream_index == videoStreamIdx)
                {
                    var sendRet = ffmpeg.avcodec_send_packet(codecCtx, pkt);
                    ffmpeg.av_packet_unref(pkt);

                    if (sendRet < 0)
                    {
                        continue;
                    }

                    while (ffmpeg.avcodec_receive_frame(codecCtx, frame) == 0)
                    {
                        var w = frame->width;
                        var h = frame->height;
                        if (w > 0 && h > 0)
                        {
                            _width = w;
                            _height = h;

                            var dstW = _targetWidth > 0 ? _targetWidth : w;
                            var dstH = _targetHeight > 0 ? _targetHeight : h;

                            EnsureBuffers(dstW, dstH);
                            EnsureSwsContext(w, h, (AVPixelFormat)frame->format, dstW, dstH);

                            if (_writeableBitmap != null && _swsContext != null)
                            {
                                for (uint i = 0; i < 8; i++)
                                {
                                    _srcData[i] = frame->data[i];
                                    _srcStride[i] = frame->linesize[i];
                                }

                                // Direct zero-copy scale into locked WriteableBitmap surface
                                using (var fb = _writeableBitmap.Lock())
                                {
                                    _dstData[0] = (byte*)fb.Address;
                                    _dstStride[0] = fb.RowBytes;
                                    for (uint i = 1; i < 8; i++)
                                    {
                                        _dstData[i] = null;
                                        _dstStride[i] = 0;
                                    }

                                    ffmpeg.sws_scale(
                                        _swsContext,
                                        _srcData,
                                        _srcStride,
                                        0,
                                        h,
                                        _dstData,
                                        _dstStride
                                    );
                                }

                                var pts = frame->pts != ffmpeg.AV_NOPTS_VALUE ? frame->pts : frame->best_effort_timestamp;
                                Interlocked.Exchange(ref _currentPts, pts);
                                Interlocked.Increment(ref _decodedFrames);
                                _hasNewFrame = true;
                            }
                        }

                        ffmpeg.av_frame_unref(frame);
                    }
                }
                else
                {
                    ffmpeg.av_packet_unref(pkt);
                }
            }

            _isConnected = false;
            ffmpeg.avcodec_free_context(&codecCtx);
            ffmpeg.avformat_close_input(&fmtCtx);
            if (_swsContext != null)
            {
                ffmpeg.sws_freeContext(_swsContext);
                _swsContext = null;
            }

            if (!_cts.IsCancellationRequested)
            {
                Thread.Sleep(3000); // 3-second backoff before reconnecting per Master Spec
            }
        }

        ffmpeg.av_frame_free(&frame);
        ffmpeg.av_packet_free(&pkt);

        if (_swsContext != null)
        {
            ffmpeg.sws_freeContext(_swsContext);
            _swsContext = null;
        }
    }

    private void RequestRender()
    {
        if (Interlocked.CompareExchange(ref _renderPending, 1, 0) == 0)
        {
            Dispatcher.UIThread.Post(() =>
            {
                _renderPending = 0;
                if (_hasNewFrame)
                {
                    _hasNewFrame = false;
                    FrameRendered?.Invoke();
                }
            }, DispatcherPriority.Render);
        }
    }

    public void Stop()
    {
        _cts.Cancel();
    }

    public void Dispose()
    {
        Stop();
        try
        {
            _workerTask?.Wait(2000);
        }
        catch
        {
            // Ignore
        }
        _cts.Dispose();
    }
}

File: cpu/CPP/src/stream_worker.cpp
#include "stream_worker.h"

#include <iostream>
#include <chrono>
#include <cstring>

static int ffmpeg_interrupt_callback(void* opaque) {
    auto* worker = static_cast<StreamWorker*>(opaque);
    return (worker && worker->isInterrupted()) ? 1 : 0;
}

StreamWorker::StreamWorker(int streamId, const std::string& rtspUrl, int targetWidth, int targetHeight, QObject* parent)
    : QThread(parent)
    , m_streamId(streamId)
    , m_rtspUrl(rtspUrl)
    , m_targetWidth(targetWidth)
    , m_targetHeight(targetHeight)
{
    setObjectName(QString("StreamWorker-%1").arg(streamId));
}

StreamWorker::~StreamWorker() {
    stopWorker();
    freeBuffers();
    freeSwsContext();
}

void StreamWorker::stopWorker() {
    m_stopRequested.store(true, std::memory_order_release);
    requestInterruption();
    if (isRunning()) {
        wait(2000);
        if (isRunning()) {
            terminate();
            wait(500);
        }
    }
}

bool StreamWorker::isInterrupted() const {
    return m_stopRequested.load(std::memory_order_relaxed) || isInterruptionRequested();
}

void StreamWorker::ensureBuffers(int width, int height) {
    int dstW = (m_targetWidth > 0) ? m_targetWidth : width;
    int dstH = (m_targetHeight > 0) ? m_targetHeight : height;
    size_t required = (size_t)dstW * dstH * 4;
    if (required > m_bufferCapacity || !m_buffers[0]) {
        freeBuffers();
        for (int i = 0; i < 3; ++i) {
            m_buffers[i] = static_cast<uint8_t*>(av_malloc(required));
            if (m_buffers[i]) {
                std::memset(m_buffers[i], 0, required);
            }
        }
        m_bufferCapacity = required;
        m_width.store(dstW, std::memory_order_release);
        m_height.store(dstH, std::memory_order_release);
    }
}

void StreamWorker::freeBuffers() {
    for (int i = 0; i < 3; ++i) {
        if (m_buffers[i]) {
            av_free(m_buffers[i]);
            m_buffers[i] = nullptr;
        }
    }
    m_bufferCapacity = 0;
}

void StreamWorker::ensureSwsContext(int width, int height, int format) {
    int dstW = (m_targetWidth > 0) ? m_targetWidth : width;
    int dstH = (m_targetHeight > 0) ? m_targetHeight : height;
    if (!m_swsCtx || m_swsWidth != width || m_swsHeight != height || m_swsFormat != format) {
        freeSwsContext();
        m_swsCtx = sws_getContext(
            width, height, static_cast<AVPixelFormat>(format),
            dstW, dstH, AV_PIX_FMT_RGB32,
            SWS_FAST_BILINEAR, nullptr, nullptr, nullptr
        );
        m_swsWidth = width;
        m_swsHeight = height;
        m_swsFormat = format;
    }
}

void StreamWorker::freeSwsContext() {
    if (m_swsCtx) {
        sws_freeContext(m_swsCtx);
        m_swsCtx = nullptr;
    }
    m_swsWidth = 0;
    m_swsHeight = 0;
    m_swsFormat = AV_PIX_FMT_NONE;
}

void StreamWorker::recordPresentedFrame(int64_t pts) {
    auto now = std::chrono::steady_clock::now();
    auto nowNs = std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count();
    int64_t prevNs = m_lastPresentedTimestampNs.exchange(nowNs, std::memory_order_acq_rel);
    if (prevNs > 0) {
        float deltaMs = static_cast<float>(nowNs - prevNs) / 1000000.0f;
        m_lastDeltaMs.store(deltaMs, std::memory_order_relaxed);
    }
    m_paintedFrames.fetch_add(1, std::memory_order_relaxed);
}

uint8_t* StreamWorker::acquireFrame(int& outWidth, int& outHeight) {
    if (m_hasNewFrame.exchange(false, std::memory_order_acq_rel)) {
        m_consumerIndex = m_sharedIndex.exchange(m_consumerIndex, std::memory_order_acq_rel);
        recordPresentedFrame(m_currentPts.load(std::memory_order_relaxed));
    }
    outWidth = m_width.load(std::memory_order_acquire);
    outHeight = m_height.load(std::memory_order_acquire);
    return m_buffers[m_consumerIndex];
}

void StreamWorker::run() {
    AVPacket* pkt = av_packet_alloc();
    AVFrame* frame = av_frame_alloc();

    while (!isInterrupted()) {
        AVFormatContext* fmtCtx = avformat_alloc_context();
        if (!fmtCtx) {
            msleep(500);
            continue;
        }

        fmtCtx->interrupt_callback.callback = ffmpeg_interrupt_callback;
        fmtCtx->interrupt_callback.opaque = this;

        AVDictionary* opts = nullptr;
        av_dict_set(&opts, "rtsp_transport", "tcp", 0);
        av_dict_set(&opts, "stimeout", "5000000", 0);      // 5 sec timeout (in us)
        av_dict_set(&opts, "max_delay", "500000", 0);       // 500 ms max delay (in us)
        av_dict_set(&opts, "buffer_size", "4194304", 0);    // 4MB socket buffer

        int ret = avformat_open_input(&fmtCtx, m_rtspUrl.c_str(), nullptr, &opts);
        av_dict_free(&opts);

        if (ret < 0) {
            m_isConnected.store(false, std::memory_order_release);
            avformat_close_input(&fmtCtx);
            if (!isInterrupted()) {
                msleep(3000);
            }
            continue;
        }

        if (avformat_find_stream_info(fmtCtx, nullptr) < 0) {
            m_isConnected.store(false, std::memory_order_release);
            avformat_close_input(&fmtCtx);
            if (!isInterrupted()) {
                msleep(3000);
            }
            continue;
        }

        int videoStreamIdx = av_find_best_stream(fmtCtx, AVMEDIA_TYPE_VIDEO, -1, -1, nullptr, 0);
        if (videoStreamIdx < 0) {
            m_isConnected.store(false, std::memory_order_release);
            avformat_close_input(&fmtCtx);
            if (!isInterrupted()) {
                msleep(3000);
            }
            continue;
        }

        AVCodecParameters* codecPar = fmtCtx->streams[videoStreamIdx]->codecpar;
        const AVCodec* codec = avcodec_find_decoder(codecPar->codec_id);
        if (!codec) {
            std::cerr << "[Stream " << m_streamId << "] Codec not found for ID: " << codecPar->codec_id << std::endl;
            m_isConnected.store(false, std::memory_order_release);
            avformat_close_input(&fmtCtx);
            if (!isInterrupted()) {
                msleep(3000);
            }
            continue;
        }

        AVCodecContext* codecCtx = avcodec_alloc_context3(codec);
        if (!codecCtx) {
            m_isConnected.store(false, std::memory_order_release);
            avformat_close_input(&fmtCtx);
            if (!isInterrupted()) {
                msleep(3000);
            }
            continue;
        }

        if (avcodec_parameters_to_context(codecCtx, codecPar) < 0) {
            m_isConnected.store(false, std::memory_order_release);
            avcodec_free_context(&codecCtx);
            avformat_close_input(&fmtCtx);
            if (!isInterrupted()) {
                msleep(3000);
            }
            continue;
        }

        // Software decoding constraints & optimizations
        codecCtx->thread_count = 1; // Dedicated QThread per stream worker
        codecCtx->flags |= AV_CODEC_FLAG_LOW_DELAY;
        codecCtx->flags2 |= AV_CODEC_FLAG2_FAST;

        if (avcodec_open2(codecCtx, codec, nullptr) < 0) {
            m_isConnected.store(false, std::memory_order_release);
            avcodec_free_context(&codecCtx);
            avformat_close_input(&fmtCtx);
            if (!isInterrupted()) {
                msleep(3000);
            }
            continue;
        }

        m_isConnected.store(true, std::memory_order_release);

        // Demuxing and decoding loop
        while (!isInterrupted()) {
            ret = av_read_frame(fmtCtx, pkt);
            if (ret < 0) {
                // Connection lost or EOF; break to reconnect
                break;
            }

            if (pkt->stream_index == videoStreamIdx) {
                int sendRet = avcodec_send_packet(codecCtx, pkt);
                av_packet_unref(pkt);

                if (sendRet < 0) {
                    continue;
                }

                while (avcodec_receive_frame(codecCtx, frame) == 0) {
                    int w = frame->width;
                    int h = frame->height;
                    if (w > 0 && h > 0) {
                        m_sourceWidth.store(w, std::memory_order_relaxed);
                        m_sourceHeight.store(h, std::memory_order_relaxed);
                        ensureBuffers(w, h);
                        ensureSwsContext(w, h, frame->format);

                        if (m_buffers[m_producerIndex] && m_swsCtx) {
                            int dstW = (m_targetWidth > 0) ? m_targetWidth : w;
                            uint8_t* dstData[4] = { m_buffers[m_producerIndex], nullptr, nullptr, nullptr };
                            int dstLinesize[4] = { dstW * 4, 0, 0, 0 };

                            sws_scale(
                                m_swsCtx,
                                frame->data,
                                frame->linesize,
                                0,
                                h,
                                dstData,
                                dstLinesize
                            );

                            // Lock-free triple buffer publish
                            int64_t pts = (frame->pts != AV_NOPTS_VALUE) ? frame->pts : frame->best_effort_timestamp;
                            m_currentPts.store(pts, std::memory_order_release);
                            m_producerIndex = m_sharedIndex.exchange(m_producerIndex, std::memory_order_acq_rel);
                            m_hasNewFrame.store(true, std::memory_order_release);
                            m_decodedFrames.fetch_add(1, std::memory_order_relaxed);
                        }
                    }
                    av_frame_unref(frame);
                }
            } else {
                av_packet_unref(pkt);
            }
        }

        m_isConnected.store(false, std::memory_order_release);
        avcodec_free_context(&codecCtx);
        avformat_close_input(&fmtCtx);
        freeSwsContext();

        if (!isInterrupted()) {
            msleep(3000); // 3-second backoff before reconnecting per Master Spec
        }
    }

    av_frame_free(&frame);
    av_packet_free(&pkt);
}
